arbstate/espresso_streamer.go

```diff
@@ -0,0 +1,355 @@
+	delayedMessageCount = new(big.Int)
+
+	// Use method.Outputs.Unpack to decode the result
+	decodedResults, err := method.Outputs.Unpack(result)
+	if err != nil {
+		return false, fmt.Errorf("failed to unpack result: %w", err)
+	}
+
+	// Ensure the decoded result is of the expected type and length
+	if len(decodedResults) != 1 {
+		return false, fmt.Errorf("unexpected number of results: expected 1, got %d", len(decodedResults))
+	}
+
+	// Assert the type of the decoded result
+	delayedMessageCount, ok = decodedResults[0].(*big.Int)
+	if !ok {
+		return false, fmt.Errorf("failed to assert result as *big.Int")
+	}
+
+	// Log the delayed message count
+	log.Info("Delayed message count", "delayedMessageCount", delayedMessageCount)
+
+	// If the delayed message count is less than the finalized delayed message read count,
+	// the delayed message count is not finalized yet
+	if msg.DelayedMessagesRead < delayedMessageCount.Uint64() {
+		return false, fmt.Errorf("delayed message count is less than the finalized delayed message read count")
+	}
+
+	return true, nil
+}
+
+/* Verify the attestation quote */
+func (s *EspressoStreamer) verifyAttestationQuote(ctx context.Context, attestation []byte, userDataHash [32]byte) error {
+
+	method, ok := s.espressoTEEVerifierABI.Methods["verify"]
+	if !ok {
+		return fmt.Errorf("verify method not found")
+	}
+
+	calldata, err := method.Inputs.Pack(
+		attestation,
+		userDataHash,
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to pack calldata: %w", err)
+	}
+	fullCalldata := append([]byte{}, method.ID...)
+	fullCalldata = append(fullCalldata, calldata...)
+
+	_, err = s.l1Reader.Client().CallContract(
+		ctx,
+		ethereum.CallMsg{
+			To:   &s.espressoTEEVerifierAddr,
+			Data: fullCalldata,
+		}, nil,
+	)
+	if err != nil {
+		return fmt.Errorf("call to the contract failed: %w", err)
+	}
+
+	return nil
+}
+
+/**
+* Create a queue of messages from the hotshot to be processed by the node
+* It will sort the messages by the message index
+* and store the messages in `messagesWithMetadata` queue
+ */
+func (s *EspressoStreamer) queueMessagesFromHotshot(ctx context.Context) error {
+	if s.nextHotshotBlockNum == 0 {
+		// We dont need to check majority here  because when we eventually go
+		// to fetch a block at a certain height,
+		// we will check that a quorum of nodes agree on the block at that height,
+		// which wouldn't be possible if we were somehow given a height
+		// that wasn't finalized at all
+		latestBlock, err := s.espressoClients[0].FetchLatestBlockHeight(ctx)
+		if err != nil {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return err
+		}
+		log.Info("Started node at the latest hotshot block", "block number", latestBlock)
+		s.nextHotshotBlockNum = latestBlock
+	}
+
+	nextHotshotBlockNum := s.nextHotshotBlockNum
+
+	// TODO: should support multiple clients
+	arbTxns, err := s.espressoClients[0].FetchTransactionsInBlock(ctx, nextHotshotBlockNum, s.namespace)
```

Comment:
I am thinking we should support the multiple client in `espresso-sequencer-go`. That should help us a lot.

Reply:
https://github.com/EspressoSystems/espresso-sequencer-go/pull/64

Reply:
Thanks @ImJeremyHe, I will take a look at that PR